### ISSUE TITLE
Can we pass LMS authentication through to an embedded H client?

### ISSUE THREAD

**Issue by dwhly** [hypothesis/lms#855]
Right now, our LMS capability works with documents hosted inside the LMS or uses via to proxy documents from elsewhere.  But if we can also bring that classroom annotation experience to documents that are hosted in on major platforms, aggregators or publishers then that is a game changer.

Why?

Two reasons:
- First, because teachers wouldn't have to download those documents from those platforms and then upload them to teach with them.  (Hassle).
- Second, because often doing even that is a violation of copyright and is either implicitly or explicitly prohibited by both the TOS of the platform and often by the formal policies of the institution (Legal).  That doesn't mean that teachers don't do it, but it means that since it falls outside policy that teaching w/ annotation on these documents it can't be mainstreamed.

As we begin partnering with these organizations in earnest, enabling them to embed H in a way that also means it can be used as the scaffolding to bring an LMS authenticated context to their platforms is a big advantage.

Proposal:  Find a way to pass LMS credentials in to our H app embedded elsewhere.

**Comment by seanh** [comment]
We already have two working solutions that I think will suffice for this: the mechanism that eLife uses and the (better) mechanism that our LMS app uses. Neither solution is eLife-, LMS- or Via-specific, and should be easy to adapt to any platform.

The mechanism that eLife uses is documented here: https://h.readthedocs.io/projects/client/en/latest/publishers/config/

We don't have any public docs for LMS's mechanism (since it's only used by our own LMS app, so far) but I'd recommend that we go with that solution for future integrations, it's better.

**Comment by dwhly** [comment]
@seanh I *think* what this means is that it could work if the content was embedded in a frame that the LMS controlled, and within that frame H was embedded on the content.  Is that correct? Because the message is passing w/in established frame-to-frame channels.

But my understanding is that today, we could not pass LMS auth through a URL (and not a "via" URL because often the content will be paywalled) to a page that has H embedded in it)?

If all this is true, then for the initial MVP it would mean the content would always be within an LMS-controlled frame.

So, a final question is:  If we can't (or don't want to) always guarantee that the platform content will be within an iframe, is it possible to also find a way to pass authenticated context through URL params alone?

**Comment by seanh** [comment]
The mechanism that the LMS app uses allows messages to be passed between any two tabs, windows or frames in the same browser -- the content doesn't have to be in an iframe. The publisher will have to do some work, adding some code to their pages to establish the page <-> Hypothesis client trust relationship and to respond to messages. But I think it should be simple enough

**Comment by dwhly** [comment]
Right.. but the LMS app and the content have to both be active at the same time inside the browser context-- even if in different tabs.  My point is that traditionally w/ a URL, you leave the context of the page you were on and pass completely to the context of the page you arrive at.  What we have at the moment would not transmit through the URL alone, right?

**Comment by seanh** [comment]
So these content pages have the client embedded? And the content pages are owned and controlled by the partner?

If so they could add the necessary code to the content pages to configure the client, using either the eLife or the LMS configuration mechanism.

So there'd seem to be no need for the original linking page to put anything in the URL -- it would be a feature of the content page itself, that it configures the client in a certain way, and it wouldn't matter how you got to the content page (by following a certain link or by navigating to it some other way).

Maybe I'm missing something..

**Comment by dwhly** [comment]
@seanh and I just had a very helpful conversation which gave me insights that I did not yet have.
- The reason why we can't stream PDFs into our app now from Google drive or other services and need to use via is because there is a kind of CORS protection that disallows downloading, but for which the proxy still works.  (Which is nuts btw).
- If we want to solve this problem w/ specific content systems then we should likely try to get a way to disable the CORS protection or include the necessary CORS header with those specific systems.

Sean provided this link which gives a nice background: https://github.com/mozilla/pdf.js/wiki/Frequently-Asked-Questions#can-i-load-a-pdf-from-another-server-cross-domain-request

**Comment by dwhly** [comment]
I am reopening this *outside* the context of PDF.  As we are hearing more and more, publishers want to teach interactive HTML versions of textbooks.  The PDF is the poor stepchild of this experience, and often specifically discouraged by the publisher.

How can we pass an authenticated LMS session into the publisher's HTML version of a textbook *without* using via (which strips away the IP or user authentication that publisher's rely upon to know whether the use is authorized)?

Assuming that some form of Hypothesis would need to be embedded by the publisher on their live HTML textbook.  Something that could received LMS credentials and properly overlay annotation there.

To discuss.

**Comment by seanh** [comment]
Does this differ from what we do for eLife? That doesn't involve Via or PDFs

**Comment by dwhly** [comment]
It differs in that we're not passing auth into the eLife site from an external context (aka an LMS session).

**Comment by seanh** [comment]
Are these HTML textbooks are being used _inside the LMS_? Not in a different tab/window/site but inside LMS's, like with our LMS assignments currently?

**Comment by dwhly** [comment]
I think right now they link from the LMS to the live HTML page

**Comment by seanh** [comment]
Do the live HTML pages already have the Hypothesis client embedded in them?

**Comment by dwhly** [comment]
No they do not.  But as part of this we presume we would be.

**Comment by klemay** [comment]
An example workflow using this online interactive text book (note that this book has H embedded already!): https://www.accessengineeringlibrary.com/content/book/9780071834087

1. A school works out a deal with McGraw-Hill to purchase access for students. 
2. A teacher decides to assign this book (or selections from it)
3. In the LMS, the teacher provides this link to students: https://www.accessengineeringlibrary.com/content/book/9780071834087
4. Students go to https://www.accessengineeringlibrary.com/content/book/9780071834087 and sign in to their (AccessEngineering) account to access the text. (This does not happen in the LMS - at this point, students are _leaving_ the LMS to view the text book). 
5. McGraw-Hill has embedded Hypothesis over this book. Students can annotate with their LMS-provisioned accounts

What's happening in this hypothetical scenario is that students are using their LMS-provisioned H accounts outside of the LMS. (Might be worth linking to https://github.com/hypothesis/lms/issues/270 here?)

**Comment by seanh** [comment]
The existing workflow if you want to annotate an HTML page, Canvas file, or Google Drive file is:

1. Teacher creates Hypothesis assignment in LMS
2. Hypothesis LMS shows our "document selection" UI
3. Teacher chooses document from Canvas files or Google Drive, or pastes in a URL
4. Students launch the assignment in Canvas

I think we should consider reusing that workflow here. So in step (3), we would add McGraw-Hill ebooks as a possible document source.

This wouldn't necessarily have to work like it does for Canvas and Google Drive files where our app shows the teacher a list of files and they click on one. If that's difficult/impossible/undesirable for some reason, then maybe when the teacher selects the "McGraw Hill ebook" option when creating an assignment we just show them a dialog that says "Paste your McGraw Hill ebook" URL here.

As a matter of fact ... We already _do_ provide a dialog to paste in _any_ URL so maybe we just need to make McGraw-Hill ebooks work with that.

The problem with the current McGraw-Hill workflow (described by Katelyn above) seems to be that if users are following a link _out of_ the LMS then they're no longer in the context of the LMS, no longer in the context of an LTI launch, so it's difficult to see how LMS features such as grading could work if the assignment is no longer in the LMS.

Additionally, it's hard to see how a single static link copied from McGraw-Hill's website by the teacher and pasted into the LMS, can possibly log different students into difference accounts when clicked on, if it's _a single static link_ (the same URL and query params) for everyone.

It seems to me that the Hypothesis LMS app itself would need to be inserted somehow into the process of selecting a McGraw-HIll ebook and then generating an LTI launch URL for it ... But that's what we currently do for Canvas and Google Drive files, so that leads back to the idea of just adding McGraw-Hill as an option into our existing workflow.

**Comment by robertknight** [comment]
I agree with Sean. If we can integrate partner publishers' content into the existing LMS assignment creation flow for Hypothesis - either as a new content source, in addition to the three that we currently have), or by automagically recognizing certain content URLs in the existing "Enter URL" dialog, that would provide a workable solution.

From a technical point of view, everything would work the same as it does for an HTML assignment, except that we wouldn't serve the content through Via but would instead serve the content directly from the publisher, who would need to configure the Hypothesis client to fetch its configuration from the LMS app in the same way that Via does. Since the publisher's own URL is loaded directly inside an iframe by our LMS app, IP information or other authentication methods would be available.

In addition to the LMS user's identity, an LTI launch also gives us additional context that we need:

- The current group
- The users' roles (student, teacher etc.) within the LMS
- Links to services provided by the LMS to LTI tools for accessing/submitted grades etc.

**Comment by judell** [comment]
This makes a ton of sense. Thanks!

**Comment by dwhly** [comment]
> If we can integrate partner publishers' content into the existing LMS assignment creation flow for Hypothesis - either as a new content source, in addition to the three that we currently have), or by automagically recognizing certain content URLs in the existing "Enter URL" dialog, that would provide a workable solution.

This feels like a very practical first step.  I imagine we'd eventually want ways that annotation could be added to any material by default w/o first going through the H workflow, but those are larger structural objectives that we could achieve through partnerships and broader integration.

> we wouldn't serve the content through Via but would instead serve the content directly from the publisher, who would need to configure the Hypothesis client to fetch its configuration from the LMS app in the same way that Via does. Since the publisher's own URL is loaded directly inside an iframe by our LMS app, IP information or other authentication methods would be available.

It's possible to launch in a separate window even if we're within the LMS app (i.e. full frame), right?

**Comment by robertknight** [comment]
> It's possible to launch in a separate window even if we're within the LMS app (i.e. full frame), right?

Yes, providing that the LMS offers this option for LTI tools. Canvas and Moodle do. I haven't checked other LMSes.

**Comment by seanh** [comment]
One thing I noticed about <https://www.accessengineeringlibrary.com/content/book/9780071834087>: this is a multi-page ebook. You click to open chapters, and each chapter is its own page/URL. This breaks the LMS app: https://github.com/hypothesis/lms/issues/451 We'll have to add support for navigating to get these ebooks working

**Comment by klemay** [comment]
@seanh yep, and the AccessEngineering text isn't unique - most of these interactive text books are multi-page.

**Comment by robertknight** [comment]
This issue describes one aspect of a solution to the problem of "how can we enable non-public content to be used with the LMS app" with a subtext of "how can we reduce the amount of work that needs to be done for each integration"?

We've since built integrations with VitalSource and JSTOR which solve this for specific content partners. One conclusion I draw from both is that it is too early to say what a generic solutions looks like. I think the only way we're going to tackle this is to build integrations with specific content partners, and aim to generalize as we go.

Some of the things that differ between non-public content providers, that we need to handle in order to deliver a useful end-user experience:

1. How authentication works (JSTOR are used to using IP-based auth, with proxy servers to facilitate off-campus access. VitalSource supports individual user accounts, but AIUI we learned from trials  that we really need to use automatic sign-on based on the user's LTI identity)
2. How the content is served (JSTOR can give us a signed PDF URL, VitalSource use their own reader), which affects how we add Hypothesis
3. Quirks of the content itself (VitalSource's custom ebook reader, JSTOR's OCR issues etc.), that we need to adapt the client to
4. How we check that an article ID or link is one of the supported ones for a particular content provider. The current UI involves getting the user to paste a link from which we extract the article ID (or a canonical link) and then use a partner's API to present a thumbnail and metadata to indicate they chose the correct thing
